File: core-backend/handlers/tts_fast.go
package handlers

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"core-backend/client"
	"core-backend/db"
	"core-backend/middleware"
	"core-backend/state"

	"github.com/bytedance/sonic"
	"github.com/google/uuid"
)

// TTSFastHandler xử lý các yêu cầu tổng hợp tiếng nói tốc độ cao (Fast TTS Engine).
type TTSFastHandler struct {
	TTSClient *client.CoreTTSClient
	VoiceMap  map[string]string
}

// NewTTSFastHandler khởi tạo TTSFastHandler cùng bảng ánh xạ tên giọng thân thiện sang mã voice code hệ thống.
func NewTTSFastHandler(ttsClient *client.CoreTTSClient) *TTSFastHandler {
	return &TTSFastHandler{
		TTSClient: ttsClient,
		VoiceMap: map[string]string{
			"Hoài Mỹ (Nữ)":       "vi-VN-HoaiMyNeural",
			"Nam Minh (Nam)":     "vi-VN-NamMinhNeural",
			"Hoài Mỹ":            "vi-VN-HoaiMyNeural",
			"Nam Minh":           "vi-VN-NamMinhNeural",
			"vi-VN-HoaiMyNeural": "vi-VN-HoaiMyNeural",
			"vi-VN-NamMinhNeural": "vi-VN-NamMinhNeural",
		},
	}
}

// FastSynthesizeRequest cấu trúc thông số yêu cầu tổng hợp Fast TTS.
type FastSynthesizeRequest struct {
	Text        string   `json:"text"`
	Voice       string   `json:"voice"`
	Speed       float64  `json:"speed"`
	JobID       *string  `json:"job_id"`
	ChunkIndex  *int     `json:"chunk_index"`
	TotalChunks *int     `json:"total_chunks"`
	TaskID      *string  `json:"task_id"`
}

// GetVoices lấy danh sách các giọng Fast TTS hỗ trợ.
func (h *TTSFastHandler) GetVoices(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	_ = sonic.ConfigDefault.NewEncoder(w).Encode([]string{"Hoài Mỹ (Nữ)", "Nam Minh (Nam)"})
}

// Synthesize tổng hợp tiếng nói Fast TTS cực nhanh trong background và trả về audio MP3.
func (h *TTSFastHandler) Synthesize(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	user, ok := middleware.GetCurrentUser(r)
	if !ok {
		w.WriteHeader(http.StatusUnauthorized)
		_ = sonic.ConfigDefault.NewEncoder(w).Encode(map[string]string{"detail": "Not authenticated"})
		return
	}

	var req FastSynthesizeRequest
	if err := sonic.ConfigDefault.NewDecoder(r.Body).Decode(&req); err != nil || strings.TrimSpace(req.Text) == "" {
		w.WriteHeader(http.StatusBadRequest)
		_ = sonic.ConfigDefault.NewEncoder(w).Encode(map[string]string{"detail": "Văn bản trống"})
		return
	}

	if req.Voice == "" {
		req.Voice = "Hoài Mỹ (Nữ)"
	}
	if req.Speed <= 0 {
		req.Speed = 1.0
	}

	voiceCode, exists := h.VoiceMap[req.Voice]
	if !exists {
		voiceCode = req.Voice
	}

	taskID := uuid.NewString()
	if req.TaskID != nil && *req.TaskID != "" {
		taskID = *req.TaskID
	}

	taskItem := state.GlobalTaskManager.GetOrCreate(taskID)

	if req.JobID != nil && *req.JobID != "" && req.ChunkIndex != nil && req.TotalChunks != nil {
		_ = db.RegisterJobAndChunk(context.Background(), user.ID, *req.JobID, "fast", req.Voice, req.Speed, *req.TotalChunks, taskID, *req.ChunkIndex, req.Text)
	}

	go func() {
		bgCtx := context.Background()
		audioBytes, err := h.TTSClient.Synthesize(req.Text, voiceCode, req.Speed, "fast")
		if err != nil {
			taskItem.Notify(state.TaskUpdate{
				Status:   "error",
				Progress: taskItem.Progress,
				Error:    err.Error(),
			})
			if req.JobID != nil && *req.JobID != "" {
				errMsg := err.Error()
				_ = db.UpdateChunkStatus(bgCtx, taskID, "error", nil, &errMsg)
			}
			return
		}

		taskItem.AudioMP3 = audioBytes
		taskItem.AudioWAV = audioBytes

		_ = os.MkdirAll("storage/temp", 0755)
		filePath := filepath.Join("storage/temp", fmt.Sprintf("%s.mp3", taskID))
		_ = os.WriteFile(filePath, audioBytes, 0644)

		if req.JobID != nil && *req.JobID != "" {
			_ = db.UpdateChunkStatus(bgCtx, taskID, "done", &filePath, nil)
		}

		taskItem.Notify(state.TaskUpdate{
			Status:   "done",
			Progress: 100,
		})
	}()

	_ = sonic.ConfigDefault.NewEncoder(w).Encode(map[string]string{
		"task_id": taskID,
	})
}

File: core-backend-python/api/tts_fast.py
import uuid
import time
import gc
import os
import asyncio
from fastapi import APIRouter, HTTPException, BackgroundTasks, Depends

from core.models import User
from api.auth import get_current_active_user
from core.schemas import FastTTSRequest
from core.state import CoreTTSClient, tasks_db, cleanup_tasks_db
from core.database import update_chunk_status

router = APIRouter()

FAST_VOICES = {
    "Hoài Mỹ (Nữ)": "vi-VN-HoaiMyNeural",
    "Nam Minh (Nam)": "vi-VN-NamMinhNeural",
    "Hoài Mỹ": "vi-VN-HoaiMyNeural",
    "Nam Minh": "vi-VN-NamMinhNeural",
    "vi-VN-HoaiMyNeural": "vi-VN-HoaiMyNeural",
    "vi-VN-NamMinhNeural": "vi-VN-NamMinhNeural"
}

@router.get("/voices")
def get_fast_voices(current_user: User = Depends(get_current_active_user)):
    return ["Hoài Mỹ (Nữ)", "Nam Minh (Nam)"]

@router.post("/synthesize")
async def synthesize_fast_tts(req: FastTTSRequest, background_tasks: BackgroundTasks, current_user: User = Depends(get_current_active_user)):
    if not req.text.strip():
        raise HTTPException(status_code=400, detail="Văn bản trống")
    
    cleanup_tasks_db()
    
    task_id = req.task_id if req.task_id else str(uuid.uuid4())
    job_id = req.job_id
    
    tasks_db[task_id] = {"progress": 0, "status": "processing", "audio_mp3": None, "audio": None, "cancel": False, "created_at": time.time()}
    loop = asyncio.get_running_loop()

    if job_id and req.chunk_index is not None and req.total_chunks is not None:
        from core.database import SessionLocal
        from core.models import TTSChunk
        db = SessionLocal()
        try:
            chunk = TTSChunk(
                id=task_id,
                job_id=job_id,
                chunk_index=req.chunk_index,
                text=req.text,
                status="processing"
            )
            db.add(chunk)
            db.commit()
        except Exception as e:
            db.rollback()
            print("Error creating chunk:", e)
        finally:
            db.close()

    def process_task():
        try:
            voice_code = FAST_VOICES.get(req.voice, req.voice)
            
            # Request audio generation from Core TTS Microservice (Port 8001)
            audio_bytes = CoreTTSClient.synthesize(
                text=req.text,
                voice=voice_code,
                speed=req.speed,
                engine="fast"
            )
            
            tasks_db[task_id]["audio_mp3"] = audio_bytes
            tasks_db[task_id]["audio"] = audio_bytes
            
            os.makedirs("storage/temp", exist_ok=True)
            file_path = f"storage/temp/{task_id}.mp3"
            with open(file_path, "wb") as f:
                f.write(audio_bytes)

            if job_id:
                update_chunk_status(task_id, "done", file_path)

            tasks_db[task_id]["progress"] = 100
            tasks_db[task_id]["status"] = "done"
            
            # Notify SSE Queue immediately
            if "queue" in tasks_db[task_id]:
                loop.call_soon_threadsafe(tasks_db[task_id]["queue"].put_nowait, {"status": "done", "progress": 100})
        except Exception as e:
            tasks_db[task_id]["status"] = "error"
            tasks_db[task_id]["error"] = str(e)
            if job_id:
                update_chunk_status(task_id, "error", None, str(e))
            if "queue" in tasks_db[task_id]:
                loop.call_soon_threadsafe(tasks_db[task_id]["queue"].put_nowait, {"status": "error", "error": str(e)})
        finally:
            gc.collect()

    background_tasks.add_task(process_task)
    return {"task_id": task_id}
